drivers/ve-direct.c

```diff
@@ -0,0 +1,298 @@
+/* ve-direct.c - Model specific routines for Victron Energy Direct units
+ *
+ * Copyright (C) 2017  Petr Kubanek <kubanek@fzu.cz>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ *
+ */
+
+#include "main.h"
+#include "serial.h"
+
+#define DRIVER_NAME	"Victron Energy Direct UPS and solar controller driver"
+#define DRIVER_VERSION	"0.20"
+
+#define VE_GET          7
+#define VE_SET          8
+
+/* driver description structure */
+upsdrv_info_t upsdrv_info = {
+	DRIVER_NAME,
+	DRIVER_VERSION,
+	"Petr Kubanek <kubanek@fzu.cz>",
+	DRV_EXPERIMENTAL,
+	{ NULL }
+};
+
+char line[500];
+size_t r_start = 0;
+
+void ve_copy(const char *endl)
+{
+	size_t left = (line + r_start) - endl;
+	if (left > 0)
+		memmove(line, endl, left);
+
+	r_start = left;
+}
+
+int process_text_buffer()
+{
+	char vn[50];
+
+	char *v_name;
+	char *v_value;
+
+	upsdebugx (3, "%s: buf '%s'", __func__, line);
+	// find checksum, verify checksum
+	char *checksum = strstr(line, "Checksum\t");
+	if (checksum == NULL || checksum + 9 >= line + r_start)
+		return -1;
+
+	// calculate checksum
+	char ch = 0;
+	for (v_name = line; v_name <= checksum + 8; v_name++)
+		ch += *v_name;
+	ch = ~ch + 1;
+	if (ch != checksum[9])
+	{
+		upslogx(1, "invalid checksum: %d %d", ch, checksum[9]);
+		ve_copy(checksum + 10);
+		return 0;
+	}
+
+	checksum[9] = '\0';
+
+	// parse what is in buffer..
+	v_name = line;
+	while (v_name < checksum)
+	{
+		if (v_name[0] != '\r' || v_name[1] != '\n')
+			break;
+		v_name += 2;
+		v_value = v_name;
+		while (*v_value != '\t' && *v_value != '\0')
+			v_value++;
+		if (v_value >= checksum)
+			break;
+		*v_value = '\0';
+		v_value++;
+		char *end = v_value;
+		while (*end != '\r' && *end != '\0')
+			end++;
+		if (*end != '\r')
+			break;
+		*end = '\0';
+
+		snprintf(vn, sizeof(vn), "ve-direct.%s", v_name);
```

Comment:
"ve-direct" is not a valid top-level NUT variable name. Please use the standard names defined, and discuss on nut-upsdev if you need anything not covered here: http://networkupstools.org/docs/developer-guide.chunked/apas01.html